src/contracts/optimism/CrosschainForwarderOptimism.sol

```diff
@@ -0,0 +1,76 @@
+ * @notice You can **only** execute payloads on optimism with up to prepayed gas which is specified in `enqueueL2GasPrepaid` gas.
+ * Prepaid gas is the maximum gas covered by the bridge without additional payment.
+ * @dev This executor is a generic wrapper to be used with Optimism CrossDomainMessenger (https://etherscan.io/address/0x25ace71c97b33cc4729cf772ae268934f7ab5fa1)
+ * It encodes a queue with a parameterless `execute()` with delegate calls and a specified target.
+ * This encoded abi is then relayed on the L2CrossDomainMessenger which is in turn queueing the proposal on the OPTIMISM_BRIDGE_EXECUTOR.
+ */
+contract CrosschainForwarderOptimism {
+  /**
+   * @dev The L1 Cross Domain Messenger contract sends messages from L1 to L2, and relays messages
+   * from L2 onto L1. In this contract it's used by the governance SHORT_EXECUTOR to send the encoded L2 queuing over the bridge.
+   */
+  address public constant L1_CROSS_DOMAIN_MESSENGER_ADDRESS =
+    0x25ace71c97B33Cc4729CF772ae268934F7ab5fA1;
+
+  /**
+   * @dev The optimism bridge executor is a L2 governance execution contract.
+   * This contract allows queuing of proposals by allow listed addresses (in this case the L1 short executor).
+   * https://optimistic.etherscan.io/address/0x7d9103572bE58FfE99dc390E8246f02dcAe6f611
+   */
+  address public constant OPTIMISM_BRIDGE_EXECUTOR =
+    0x7d9103572bE58FfE99dc390E8246f02dcAe6f611;
+
+  /**
+   * @dev The CTC contract is an append only log of transactions which must be applied to the rollup state.
+   * It also holds configurations like the currently prepayed amount of gas which is what this contract is utilizing.
+   * https://etherscan.io/address/0x5e4e65926ba27467555eb562121fac00d24e9dd2#code
+   */
+  ICanonicalTransactionChain public constant CANONICAL_TRANSACTION_CHAIN =
+    ICanonicalTransactionChain(0x5E4e65926BA27467555EB562121fac00D24E9dD2);
+
+  /**
+   * @dev this function will be executed once the proposal passes the mainnet vote.
+   * @param l2PayloadContract the polygon contract containing the `execute()` signature.
+   */
+  function execute(address l2PayloadContract) public {
+    address[] memory targets = new address[](1);
+    targets[0] = l2PayloadContract;
+    uint256[] memory values = new uint256[](1);
+    values[0] = 0;
+    string[] memory signatures = new string[](1);
+    signatures[0] = 'execute()';
+    bytes[] memory calldatas = new bytes[](1);
+    calldatas[0] = '';
+    bool[] memory withDelegatecalls = new bool[](1);
+    withDelegatecalls[0] = true;
+
+    bytes memory queue = abi.encodeWithSelector(
+      IL2BridgeExecutor.queue.selector,
+      targets,
+      values,
+      signatures,
+      calldatas,
+      withDelegatecalls
+    );
+    ICrossDomainMessenger(L1_CROSS_DOMAIN_MESSENGER_ADDRESS).sendMessage(
+      OPTIMISM_BRIDGE_EXECUTOR,
+      queue,
+      uint32(CANONICAL_TRANSACTION_CHAIN.enqueueL2GasPrepaid())
```

Comment:
If the execution costs more than l2GasPrepaid (1.9M at the time of writing) the proposal will fail. Should we estimate the gas cost in l2?

Reply:
it will always only be the price of queue, so should be fine. Might make sense to estimate though :+1:

Reply:
Oh, you are right. Lets estimate just in case, but I assume its under 1.9M

Reply:
checked and it's ~390k

Reply:
Awesome, thanks